### PR TITLE
Nmallick1/feedback for kusto queries

### PR DESCRIPTION
**Pull request by nmallick1**
## Overview
Adding mechanism to take feedback and use it in chatbots

## Type of change
- [X] New feature (non-breaking change which adds functionality)

### This PR:
* Adding mechanism to take feedback and use it in chatbots
* Introduces a new markdown wrapper for chatbot messages with individual code snippet copy option
* Sends resource information (e.g.. app kind) with every chat message
* Sends ARM id along with every chat message. This helps chat messages become resource aware
* For any chat message generated via fewshots, sends the corresponding feedbackids back in the final stream
* Generalizes the config structure for all chatbots allowing them to be enabled per resource type for specific users and selectively enable authorized feedback providers
* Adds methods to generate Kusto query deep link
* Adds methods to loosely check for kusto query best practices
* Enables Composite question creation based on last user intent
* Auto generate feedback explanation to facilitate search lookup with chain of though prompting
* Update logging to include chat identifiers to help with telemetry tracking later
* Adds support for both resource aware and top level chatbots along with an option to switch between them.
* Enable query assistant only for function apps.

## How Can This Be Tested? <span>&#128269;</span>
Replicate the branch and use the KQL feature
Test other copilots as well

## Checklist <span>&#9989;</span>
- [X] I have looked over the diffs.
- [X] I have run the linter to catch any errors.
- [X] There are no errors from my changes on this branch.
- [X] I have ensured that my changes support accessibility and can be accessed with the keyboard alone
- [X] I have added tests for this feature (if applicable)
- [X] I have incorporated telemetry into this code to capture and analyze relevant metrics (if applicable)
- [X] I have requested review on this PR.

## Screenshots After Fix (if appropriate):

image

image

image

image

### CODE REVIEW

File: AngularApp/projects/diagnostic-data/src/lib/utilities/kusto-utilities.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+export interface KustoQuery {
+    text: string;
+    Url: string;
+    KustoDesktopUrl: string;
+}
+
+export class KustoUtilities {
+    public static readonly KustoDesktopImage:string = `https://applensimgs.blob.core.windows.net/images/KustoDesktop1.png`;
+    public static readonly KustoWebImage:string = `https://applensimgs.blob.core.windows.net/images/Kustoweb1.png`;
```
Comment (rekhaswaminathan) [inline AngularApp/projects/diagnostic-data/src/lib/utilities/kusto-utilities.ts:11]: Is this an open endpoint?

Comment (nmallick1) [reply AngularApp/projects/diagnostic-data/src/lib/utilities/kusto-utilities.ts:11]: oh no.. these are just images for showing Run in Kusto Explorer/ Run in Kusto Web

image

**Comment by nmallick1** [comment]
> @nmallick1 The PR build is failing. Looks like it has something to do w/ the ChatUIMarkdownWrapperComponent
> 
> image

Thanks for catching this @osshad, fixed the build

File: ApplensBackend/Controllers/OpenAIController.cs [hunk 2]
```diff
@@ -148,28 +154,101 @@ public async Task<IActionResult> IsDetectorCopilotEnabled()
...
+
+            return Ok(false);
+        }
+
+        [HttpPost("saveChatFeedback")]
+        [HttpOptions("saveChatFeedback")]
+        public async Task<IActionResult> SaveChatFeedback([FromBody] ChatFeedback feedbackPayload)
+        {
+            var userAlias = Utilities.GetUserIdFromToken(Request.Headers.Authorization).Split(new char[] { '@' }).FirstOrDefault();
+            if (feedbackPayload.SubmittedBy?.Equals(userAlias, StringComparison.OrdinalIgnoreCase) == false)
```
Comment (osshad) [inline ApplensBackend/Controllers/OpenAIController.cs:200]: 👍

**Review by osshad — COMMENTED** [review]
Great job, Nazeef!! I was unable to submit my feedback. I guess I'm missing some settings on my end.

image


Just a UI nitpick: I think we can keep the `Submit Feedback` button disabled even when users enter empty/whitespace characters with no text. That way, they won't have to hit the `saveChatFeedback` endpoint unnecessarily 😄 

image

**Review by osshad — APPROVED** [review]

**Comment by nmallick1** [comment]
> Great job, Nazeef!! Just a small UI nitpick: I think we can keep the `Submit Feedback` button disabled even when users enter empty/whitespace characters with no text in the Expected response textbox. This would improve the feedback validation which you already have in place.
> 
> image

Thanks for testing @osshad and finding an improvement. Pushed a change.

**Review by ShekharGupta1988 — APPROVED** [review]

**Review by XiaoxuHU — APPROVED** [review]

**Review by rekhaswaminathan — APPROVED** [review]

**Pull request merged**